redaxo/src/core/lib/login/login.php

```diff
@@ -230,14 +246,58 @@ public function checkLogin()
         return $ok;
     }
 
+    public function impersonate($id)
+    {
+        if (!$this->user) {
+            throw new RuntimeException('Can not impersonate a user without valid user session.');
+        }
+        if ($this->user->getValue($this->idColumn) == $id) {
+            throw new RuntimeException('Can not impersonate the current user.');
+        }
+
+        $user = rex_sql::factory($this->DB);
+        $user->setQuery($this->userQuery, [':id' => $id]);
+
+        if (!$user->getRows()) {
+            throw new RuntimeException(sprintf('User with id "%d" not found.', $id));
+        }
+
+        $this->impersonator = $this->user;
+        $this->user = $user;
+
+        $this->setSessionVar('UID', $id);
+        $this->setSessionVar('impersonator', $this->impersonator->getValue($this->idColumn));
+    }
+
+    public function unimpersonate()
+    {
+        if (!$this->impersonator) {
+            throw new RuntimeException('There is no current impersonator.');
+        }
+
+        $this->user = $this->impersonator;
+        $this->impersonator = null;
+
+        $this->setSessionVar('UID', $this->user->getValue($this->idColumn));
+        $this->setSessionVar('impersonator', null);
+    }
+
     /**
-     * @return rex_user
+     * @return null|rex_sql
      */
     public function getUser()
     {
         return $this->user;
     }
 
+    /**
+     * @return null|rex_sql
```

Comment:
Aktuell kann es auch ein rex-user werden, oder?

Reply:
Nicht in der Basisklasse, habe daher ein `@method` in der `rex_backend_login`-Klasse ergänzt.

Ich stimme dir zu, dass das nicht so toll ist mit den unterschiedlichen Klassen, aber habe mich nur am `user` orientiert, und wollte da nun nichts am Verhalten ändern.

Allgemein soll man das Feature auch für Frontend-Logins nutzen können, deswegen habe ich es in die Basisklasse integriert.